### ISSUE TITLE
Weird behaviour when handling paths in rawhttp defined requests

### ISSUE THREAD

**Issue by SecGus** [projectdiscovery/nuclei#3475]
I have the following template which I am running through nuclei:
```
requests:
  - raw:
      - |+
        GET  HTTP/1.1
        Host: {{Hostname}}
        Accept: */*
        Accept-Language: en-GB,en;q=0.5
        Accept-Encoding: gzip, deflate
        User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:105.0) Gecko/20100101 Firefox/105.0


      
      - |+
        GET {{md5(headers)}}=1 HTTP/1.1
        Host: {{Hostname}}
        User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:105.0) Gecko/20100101 Firefox/105.0



    unsafe: true
    matchers:
      - type: dsl
        dsl:
         - "status_code_1 == 200"
    redirects: false
    req-condition: false
 ```
 
 When running the following command, it sends the following request:
 `echo "https://example.com/" | nuclei -t cpdos.yaml -dreq`

```
GET  HTTP/1.1
Host: example.com
Accept: */*
Accept-Language: en-GB,en;q=0.5
Accept-Encoding: gzip, deflate
User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:105.0) Gecko/20100101 Firefox/105.0
```

Notice the lack of a path. When I saw this, I guessed I should manually add {{Path}} as I am using rawhttp. However, when the URL naturally includes a path, such as /testing, and there is something already defined where the path would go, such as a parameter (notice in the rawhttp request the question mark), the following happens:
`echo "https://example.com/testing/" | nuclei -t cpdos.yaml -dreq`
```
requests:
  - raw:
      - |+
        GET ? HTTP/1.1
        Host: {{Hostname}}
        Accept: */*
        Accept-Language: en-GB,en;q=0.5
        Accept-Encoding: gzip, deflate
        User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:105.0) Gecko/20100101 Firefox/105.0
```

```
GET /testing/ HTTP/1.1
Host: example.com
Accept: */*
Accept-Language: en-GB,en;q=0.5
Accept-Encoding: gzip, deflate
User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:105.0) Gecko/20100101 Firefox/105.0
```

Notice the URL has been appended.

To summarise, this means that when piping standard input to the nuclei template, when a URL ends only in `/`, and not {{Path}} variable is defined in the template, it will not work, as the path in the HTTP Request is sent as blank. Whereas, if you include the {{Path}} variable in the rawHTTP request, and then pipe a URL that already has a path on it to it, the path is included twice.

This ultimately leads to there being no viable rawhttp request definition that can be applied to situations where both a path, and no path is being passed to the template.

**Comment by SecGus** [comment]
I know this is quite a confusing issue comment, I'll try and clean it up and add more information tomorrow.

**Comment by SecGus** [comment]
I'll attempt to create a table below to demonstrate the confusion:

| Input URL  | Raw Request in Template | Output Request Path  | 
| ------------- | ------------- | ------------- | 
| "https://gusralph.info/"  | `GET  HTTP/1.1` | `GET  HTTP/1.1`  |
| "https://gusralph.info/"  | `GET ?y=x HTTP/1.1` | `GET ?y=x HTTP/1.1`  |
| "https://gusralph.info/testing/" | `GET  HTTP/1.1`  | `GET  HTTP/1.1` | 
| "https://gusralph.info/testing/" | `GET ?y=x HTTP/1.1`  | `GET /testing/?y=x HTTP/1.1` | 
| "https://gusralph.info/" | `GET {{Path}} HTTP/1.1` | `GET / HTTP/1.1`  |
| "https://gusralph.info/" | `GET {{Path}}?y=x HTTP/1.1` | `GET /?y=x HTTP/1.1`  |
| "https://gusralph.info/testing/" | `GET {{Path}} HTTP/1.1` | `GET /testing/testing HTTP/1.1`  |
| "https://gusralph.info/testing/" | `GET {{Path}}?y=x HTTP/1.1` | `GET /testing/testing?y=x HTTP/1.1`  |

**Comment by tarunKoyalwar** [comment]
@SecGus  some of them seems to be wierd behaviour but most of it is known > https://github.com/projectdiscovery/nuclei/pull/3331

Some other points to Note
- By default if input contains path ex: `https://gusralph.info/testing/` then it is directly appended to raw request path at runtime so no need to add `{{Path}}`
- Empty path in case 1 & 2 is expected since it is `unsafe: true` and some exploit requires these types of situation . in unsafe:true basically whatever the input is in template it will remain the same

**Comment by tarunKoyalwar** [comment]
> I'll attempt to create a table below to demonstrate the confusion:
> 
> Input URL	Raw Request in Template	Output Request Path	@tarunKoyalwar
> "https://gusralph.info/"	`GET HTTP/1.1`	`GET HTTP/1.1`	`intended`
> "https://gusralph.info/"	`GET ?y=x HTTP/1.1`	`GET ?y=x HTTP/1.1`	`intended `
> "https://gusralph.info/testing/"	`GET HTTP/1.1`	`GET HTTP/1.1`	`wierd-behaviour`
> "https://gusralph.info/testing/"	`GET ?y=x HTTP/1.1`	`GET /testing/?y=x HTTP/1.1`	`intended`
> "https://gusralph.info/"	`GET {{Path}} HTTP/1.1`	`GET / HTTP/1.1`	`invalid-use`
> "https://gusralph.info/"	`GET {{Path}}?y=x HTTP/1.1`	`GET /?y=x HTTP/1.1`	`invalid-use`
> "https://gusralph.info/testing/"	`GET {{Path}} HTTP/1.1`	`GET /testing/testing HTTP/1.1`	`invalid-use`
> "https://gusralph.info/testing/"	`GET {{Path}}?y=x HTTP/1.1`	`GET /testing/testing?y=x HTTP/1.1`	`invalid-use`
> Notice in the latter rows of the output column, the trailing slash is also removed from the path after appending the `testing` twice.

the only edgecase/ wierd-behaviour seems to be case 3 . we will look into it but apart from that all other things are either intended or not properly used . @SecGus  refer > https://github.com/projectdiscovery/nuclei/pull/3331#issuecomment-1435263381

**Comment by tarunKoyalwar** [comment]
@SecGus  thanks for pointing out these edgecases , but do note that when `unsafe:true` urls are not normalized at all . unless the template requires such extreme cases we should stick to `unsafe: false`

**Comment by SecGus** [comment]
Hi @tarunKoyalwar,

Wouldn't it be better with some consistency, even with unsafe: true? For example, if we are going to leave the Path empty (like in 1 & 2) when no {{Path}} variable is defined, shouldn't the same happen for the rest of the cases? And that way hackers can choose if they want the Path there or not? Currently, it treats the lack of {{Path}} differently depending on the input URL, leading to quite a large restriction on diversity when considering potential URL input.

**Comment by tarunKoyalwar** [comment]
### TODO
- [ ] add docs related to such edgecases in `nuclei-docs`

**Comment by tarunKoyalwar** [comment]
> Hi @tarunKoyalwar,
> 
> Wouldn't it be better with some consistency, even with unsafe: true? For example, if we are going to leave the Path empty (like in 1 & 2) when no {{Path}} variable is defined, shouldn't the same happen for the rest of the cases? And that way hackers can choose if they want the Path there or not? Currently, it treats the lack of {{Path}} differently depending on the input URL, leading to quite a large restriction on diversity when considering potential URL input.

@SecGus , to summarize use of `{{Path}}` for handling urls is discouraged in raw requests `paths` because this directly conflicts the behaviour of automerge (i.e adjusting the templates when inputurl has paths or params `ex: https://scanme.sh/wp/?token=sdfsfsfds`

**Comment by tarunKoyalwar** [comment]
> 



> To clarify, what is the intended output from case 3? Should it be `GET /testing/ HTTP/1.1`? If so, why should this include the path but just a `/` shouldn't? To me it seems like these should all have no path in them, as they have not been defined in the template requests. Equally, by avoiding rawhttp adding the path by default, we can avoid the duplicated `/testing/testing` path that is generated in other use cases.

@SecGus  yes the intended output of case 3 is `GET /testing/ HTTP/1.1` 
as for the case if input is `https://scanme.sh/` . this is accepted risk to avoid unexpected outputs/false positives
Consider url `https://scanme.sh` and `https://scanme.sh/` <- they seem like same to user but its totally different when it comes to coding and kinda invalidates half of the templates . hence to handle such edgecases . 

if user desires `/` in raw requests it would be better to do the same in template instead of keeping it empty and nuclei will handled the extra slash 
```yaml
- raw :
     -  |
     GET /  HTTP/1.1
```
if template is in ^ format even if input is `https://scanme.sh` or `https://scanme.sh/` the request will remain same . pls do note that `GET HTTP/1.1` is invalid request and url when it comes to standard library and when `unsafe: false` it is normalized . I hope this answers your query if not do reach out at https://discord.gg/projectdiscovery we actively resolve queries/questions in our discord channel

**Comment by tarunKoyalwar** [comment]
@SecGus  we have added new template option to disable automerging of paths to address this and other similar issue which should be used when request contains immutable / unsafe paths https://github.com/projectdiscovery/nuclei/pull/3799
